### ISSUE TITLE
Is it possible to remove all require() calls in a bundled script?

### ISSUE THREAD

**Issue by guest271314** [evanw/esbuild#3947]
I'm experimenting with bundling `npm` to a standalone executable using the procedures outlined by Deno, Bun, and Node.js.

One observation of the Node.js https://nodejs.org/api/single-executable-applications.html#requireid-in-the-injected-main-script-is-not-file-based and Bun https://bun.sh/docs/bundler#executables implementations is that when the source entry points are removed from the file system the repective resulting executable throws module not found error.

For whatever reason, Node.js and Bun implementations expect the source entry point files to still be present of the file system for the executable to work.

So my idea is to bundle all dependencies into a single script that effectively refers to itself for `require()` calls, and does not try to use `require()` at all.

Is this possible using `esbuild`?

**Comment by evanw** [comment]
Sorry, I'm not quite sure what you're asking. The point of a bundler like esbuild is to bundle things, and esbuild supports bundling `require()` calls, so this is possible with esbuild in theory. However, it'll only work if your dependencies are bundler-friendly. For esbuild, being bundler-friendly roughly means:

* All `require()` calls use string literals (not run-time expressions)
* All `require()` calls resolve to JavaScript files that esbuild can understand (e.g. not to `.node` native extensions)
* The bundled code doesn't use file system APIs to load code and/or data at run-time (e.g. `fs.readFileSync`)
* The bundled code doesn't use features that require the file to be a specific location on the file system such as `__filename`

If your dependencies aren't bundler-friendly, then it may not be possible to bundle them with esbuild.

**Comment by guest271314** [comment]
> Sorry, I'm not quite sure what you're asking.

Some context: [Compiling npm to a standalone executable: Which runtime can do this out of the box; node, deno, or bun?](https://gist.github.com/guest271314/c9543a19d8ccf72881355b27d0107551).

What happens with the Bun and Node.js, respectively, implementations of their single executable applications, is that if we create the executable, then remove the original source files, the executable stops working, e.g., https://github.com/oven-sh/bun/issues/14676

```
bun-npm --help
error: Cannot find module "/home/user/bin/node_modules/npm/lib/cli/entry.js" from "/$bunfs/root/bun-npm"
      at syntaxErrorHandler (/$bunfs/root/bun-npm:287:16)
```

because of embedded `require()`s of some sort. For Node.js, because `require()` is not `require()` in the compilation process, I have to prepend this to the script bundled with `bun build` or `esbuild` for the compilation to work at all

```
node -e 'const file="bun-npm-bundle.js";const fs=require("fs");fs.writeFileSync(file, "require=require(\"node:module\").createRequire(__filename);\n" + fs.readFileSync(file, "utf8").replace("#!/usr/bin/env node", ""));'
```

because both `require("../local/path")` and `require("node:some_node_module")` are included in CommonJS source code for GitHub's `cli/npm`.

Bun's `build` from the looks of it is very similar to `esbuild` output, where we'll have something like this

```
var require_cli = __commonJS((exports, module) => {
    var __dirname = "/home/xubuntu/bin/node_modules/npm/lib";
    var validateEngines = require_validate_engines();
    var cliEntry = (init_path(), __toCommonJS(exports_path)).resolve(__dirname, "cli/entry.js");
    module.exports = (process2) => validateEngines(process2, () => __require(cliEntry));
  });
```

which is trying to reproduce something like this https://github.com/npm/cli/blob/latest/lib/cli.js

```
const validateEngines = require('./cli/validate-engines.js')
const cliEntry = require('node:path').resolve(__dirname, 'cli/entry.js')

module.exports = (process) => validateEngines(process, () => require(cliEntry))
```

So what I'm trying to figure out is how to bundle all of the source code where `require()` is used into a single bundled script that does use `require()` or try to emulate `require()` at ll. Rather just creates named functions that refer to those named functions in the same bundled script.

That is, 

```
const validateEngines = require('./cli/validate-engines.js')
const cliEntry = require('node:path').resolve(__dirname, 'cli/entry.js')
```

becomes

```
function validateEngines() { // ... }
function cliEntry() { // ... }
```

specifically for Bun and Node.js, so we can then remove the `node_modules` folder or the GitHub `cli/npm` repository, and the resulting single executable *shou;d* still work, because all of the code will be in the one file, without searching for any modules on the system.

That won't fix the fact of how Node.js and Bun have decided to implement an executable compilation that depends on the source files after compilation. Though might be a workaround. Not sure until I get to the point were I can transform `require()` calls to function declaration calls referencable within the same single file.

Make sense?

**Comment by hyrious** [comment]
I don't think it is possible to bundle `cliEntry = path.resolve(__dirname, 'cli/entry.js')` into a single executable since look, there's no real file system structure when it is distributed. A later `require(cliEntry)` is also not analyzable (i.e. bundle-able) since it uses dynamic require call, as Evan has pointed above.

I'm not sure why npm cli needs to resolve the file path before actually require it, wouldn't this code be the same?:

```js
module.exports = (process) => validateEngines(process, () => require('./cli/entry.js')) // This is a bundle-able require()
```

One possible reason behind that is npm tries to make sure there's no code > ES6 being loaded into old versions of V8 before it could validate the engine and print useful messages since otherwise there may raise a SyntaxError before it could do anything.

So basically you can try to rewrite the code like above (maybe there're other places doing the similar) and try again.

**Comment by guest271314** [comment]
@hyrious The issue is using `require()` at all in the code the creates the executable. What I'm trying to do is create a corresponding function symbol for each use of `require()` that `esbuild` currently bundles to either CommonJS or Ecmascript Module. Then I can test if the executables produced by Node.js and Bun *can* work standalone. Right now I don't think they can.

So, if I run 

```
node_modules/.bin/esbuild --bundle node_modules/npm/bin/npm-cli.js --platform=node --format=cjs --outfile=esbuild-npm-bundle.js --tree-shaking=true
```

I'll get something like the following at the bottom of the bundled script

```
// node_modules/npm/lib/cli.js
var require_cli = __commonJS({
  "node_modules/npm/lib/cli.js"(exports2, module2) {
    var validateEngines = require_validate_engines();
    var cliEntry = require("node:path").resolve("node_modules/npm/lib", "cli/entry.js");
    module2.exports = (process2) => validateEngines(process2, () => require(cliEntry));
  }
});
```

I'm trying to transform that symbol into the compete actual textual representation of `entry.js`

```
/* eslint-disable max-len */

// Separated out for easier unit testing
module.exports = async (process, validateEngines) => {
  // set it here so that regardless of what happens later, we don't
  // leak any private CLI configs to other programs
  process.title = 'npm'

  // if npm is called as "npmg" or "npm_g", then run in global mode.
  if (process.argv[1][process.argv[1].length - 1] === 'g') {
    process.argv.splice(1, 1, 'npm', '-g')
  }

  // Patch the global fs module here at the app level
  require('graceful-fs').gracefulify(require('node:fs'))

  // ...
```

where text such as 

```
require('graceful-fs').gracefulify(require('node:fs'))
```

is replaced with 
```
function require_cli() {
   function gacefulify() {
      // Text of gacefulify function
      // Repeated for any and all requires()'s here, too
   }
}
```

Looking at the `esbuild` documentation I don't see anything like that right now.

Basically I'm trying to create as runtime agnostic as possible JavaScript source code. In this case, to test Node.js and Bun's implementation when no `requires()` are present, and after the original source files are removed. I already know the case of removing the source files right now breaks Node.js and Bun's implementation of a single executable.

Deno does it differently. A folder is created in `/tmp` that is empty, and an sqlite database is created in the Deno configuration folder each time the executable is run. We can delete those directories, too, and the executable still works, after being written to an external USB, and written back to a filesystem.

**Comment by guest271314** [comment]
I think all of the `require("node:some_node_module")` references might be alright for this case.

If this is not possible or within the scope of `esbuild`, kindly just say so, and I'll try to figure something else out to continue testing this domain.

Thanks.
